### ISSUE TITLE
Predictions not consistent after loading the AutoML model from directory. 

### ISSUE THREAD

**Issue by alitirmizi23** [mljar/mljar-supervised#471]
I created an AutoML model using the "Compete" mode. When using the .predict() method after loading the model using "results_path", it gives totally different results on the same test data, compared to when using .predict() directly after training the model. 

F1 metric on test data using .predict() right after training:
![before](https://user-images.githubusercontent.com/44982627/134116014-747b70bb-f98e-4402-ae3b-0eee7619a455.png)

F1 metric on the same test data using .predict() after loading from directory:
![after](https://user-images.githubusercontent.com/44982627/134116078-c639b044-6b58-44d4-bc7a-f9ac829e0bf7.png)

Am I missing something here? Note that the best model was "Ensemble_stacked".

**Comment by pplonski** [comment]
Hi, I will need code and data to reproduce this issue. Without it is hard to say if it is a bug.

**Comment by alitirmizi23** [comment]
> Hi, I will need code and data to reproduce this issue. Without it is hard to say if it is a bug.

I am using data that's sensitive and would not be able to share that. I can share the code though..
Also, I had to add "LightGBM_18" to load_on_predict list in the params.json for the .predict() method to work after loading from directory. 

`automl = AutoML(
#     ml_task='multiclass_classification',
                mode='Compete',
                algorithms=['Baseline','CatBoost', 'LightGBM', 'Xgboost'],
#                eval_metric='f1',
                stack_models=True,
                total_time_limit= 7200,
                model_time_limit=None,
    
        
                validation_strategy=            {
                "validation_type": "kfold",
                "k_folds": 3,
                "shuffle": True,
                "stratify": True,
                "random_seed": 123
            }
               )

automl.fit(X_train_count_final, pd.Series(y_train.values[:,-1]),)

automl = AutoML(results_path='AutoML_1')

from sklearn.metrics import f1_score,confusion_matrix,balanced_accuracy_score,accuracy_score

f1_score(y_test,automl.predict(X_test_count_final),average=None)`

**Comment by alitirmizi23** [comment]
> Hi, I will need code and data to reproduce this issue. Without it is hard to say if it is a bug.

Could you maybe try to do on some dummy data yourself? This is strange behavior. Wish I could provide my data but its just not an option..

**Comment by pplonski** [comment]
Thank you for code. I will try with dummy data. How many classes you had in your data?

Regarding loading the models. I fixed the bug yesterday and updated code is in the master branch.

**Comment by alitirmizi23** [comment]
I had 8 classes in the data (multiclass problem). I'm attaching the params.json (converted to csv) for you to see how many models it saved and loaded. I think 56 (looking at the folder numbers inside the load directory). 
[params.csv](https://github.com/mljar/mljar-supervised/files/7201152/params.csv)

**Comment by alitirmizi23** [comment]
> I had 8 classes in the data (multiclass problem). I'm attaching the params.json (converted to csv) for you to see how many models it saved and loaded. I think 56 (looking at the folder numbers inside the load directory).
> [params.csv](https://github.com/mljar/mljar-supervised/files/7201152/params.csv)

Also, as noted before, I added 'LightGBM_18' manually to the file - maybe that causes predictions to deviate? I will try to update the project code from master after bug fix, and see if that helps too in the meantime

**Comment by pplonski** [comment]
@alitirmizi23 I've tried with dummy data:

```py
import os
import unittest
import tempfile
import json
import numpy as np
import pandas as pd
import shutil
from sklearn import datasets
from sklearn.model_selection import train_test_split

from supervised import AutoML


class AutoMLPredictionAfterLoadTest(unittest.TestCase):

    automl_dir = "automl_testing"

    def tearDown(self):
        shutil.rmtree(self.automl_dir, ignore_errors=True)

    def test_integration(self):
        a = AutoML(
            results_path=self.automl_dir,
            mode='Compete',
            algorithms=['Baseline','CatBoost', 'LightGBM', 'Xgboost'],
            stack_models=True,
            total_time_limit= 1200,
            validation_strategy={
                "validation_type": "kfold",
                "k_folds": 3,
                "shuffle": True,
                "stratify": True,
                "random_seed": 123
            }
        )

        X, y = datasets.make_classification(
            n_samples=1000,
            n_features=30,
            n_informative=29,
            n_redundant=1,
            n_classes=8,
            n_clusters_per_class=3,
            n_repeated=0,
            shuffle=False,
            random_state=0,
        )
        X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8)

        a.fit(X_train, y_train)
        p = a.predict_all(X_test)
        print(p)

        a2 = AutoML(results_path=self.automl_dir)
        p2 = a2.predict_all(X_test)
        print(p2)
```

I got two identical predictions. The output:
```
The task is multiclass_classification with evaluation metric logloss
AutoML will use algorithms: ['Baseline', 'CatBoost', 'LightGBM', 'Xgboost']
AutoML will stack models
AutoML will ensemble available models
AutoML steps: ['simple_algorithms', 'default_algorithms', 'not_so_random', 'golden_features', 'kmeans_features', 'insert_random_feature', 'features_selection', 'hill_climbing_1', 'hill_climbing_2', 'boost_on_errors', 'ensemble', 'stack', 'ensemble_stacked']
* Step simple_algorithms will try to check up to 1 model
1_Baseline logloss 2.079179 trained in 0.53 seconds
* Step default_algorithms will try to check up to 3 models
2_Default_LightGBM logloss 1.780787 trained in 9.76 seconds
3_Default_Xgboost logloss 1.778642 trained in 8.37 seconds
4_Default_CatBoost logloss 1.594174 trained in 4.49 seconds
* Step not_so_random will try to check up to 27 models
14_LightGBM logloss 1.733766 trained in 5.11 seconds
5_Xgboost logloss 1.747981 trained in 7.27 seconds
23_CatBoost logloss 1.683927 trained in 1.87 seconds
15_LightGBM logloss 1.831602 trained in 8.02 seconds
6_Xgboost logloss 1.806162 trained in 7.7 seconds
24_CatBoost logloss 1.594199 trained in 6.73 seconds
16_LightGBM logloss 1.75559 trained in 6.3 seconds
7_Xgboost logloss 1.801397 trained in 6.7 seconds
25_CatBoost logloss 1.569588 trained in 13.49 seconds
17_LightGBM logloss 1.780325 trained in 8.23 seconds
8_Xgboost logloss 2.079141 trained in 4.9 seconds
26_CatBoost logloss 1.564274 trained in 19.26 seconds
18_LightGBM logloss 1.766689 trained in 5.49 seconds
9_Xgboost logloss 1.872442 trained in 7.96 seconds
27_CatBoost logloss 1.592945 trained in 6.62 seconds
19_LightGBM logloss 1.71987 trained in 7.65 seconds
10_Xgboost logloss 1.835022 trained in 7.2 seconds
28_CatBoost logloss 1.707934 trained in 2.42 seconds
20_LightGBM logloss 1.726679 trained in 6.52 seconds
11_Xgboost logloss 1.690941 trained in 11.09 seconds
29_CatBoost logloss 1.588272 trained in 10.81 seconds
21_LightGBM logloss 1.816244 trained in 7.14 seconds
12_Xgboost logloss 1.866249 trained in 7.14 seconds
30_CatBoost logloss 1.664913 trained in 3.32 seconds
22_LightGBM logloss 1.747241 trained in 7.18 seconds
13_Xgboost logloss 1.704708 trained in 11.82 seconds
31_CatBoost logloss 1.538194 trained in 30.51 seconds
* Step golden_features will try to check up to 3 models
None 10
Add Golden Feature: feature_13_ratio_feature_4
Add Golden Feature: feature_4_ratio_feature_7
Add Golden Feature: feature_21_multiply_feature_7
Add Golden Feature: feature_20_ratio_feature_21
Add Golden Feature: feature_2_diff_feature_27
Add Golden Feature: feature_27_ratio_feature_7
Add Golden Feature: feature_16_ratio_feature_25
Add Golden Feature: feature_17_ratio_feature_19
Add Golden Feature: feature_4_diff_feature_27
Add Golden Feature: feature_21_multiply_feature_3
Created 10 Golden Features in 3.81 seconds.
31_CatBoost_GoldenFeatures logloss 1.62384 trained in 20.86 seconds
26_CatBoost_GoldenFeatures logloss 1.619728 trained in 12.03 seconds
25_CatBoost_GoldenFeatures logloss 1.63555 trained in 12.23 seconds
* Step kmeans_features will try to check up to 3 models
31_CatBoost_KMeansFeatures logloss 1.596976 trained in 25.78 seconds
26_CatBoost_KMeansFeatures logloss 1.605875 trained in 17.65 seconds
25_CatBoost_KMeansFeatures logloss 1.614677 trained in 19.44 seconds
* Step insert_random_feature will try to check up to 1 model
31_CatBoost_RandomFeature logloss 1.552775 trained in 22.69 seconds
Drop features ['random_feature']
Skip features_selection because no parameters were generated.
* Step hill_climbing_1 will try to check up to 11 models
32_CatBoost logloss 1.562252 trained in 8.45 seconds
33_CatBoost logloss 1.567423 trained in 8.37 seconds
34_CatBoost logloss 1.60311 trained in 6.77 seconds
35_Xgboost logloss 1.704463 trained in 8.97 seconds
36_Xgboost logloss 1.70719 trained in 10.04 seconds
37_LightGBM logloss 1.71987 trained in 7.73 seconds
38_LightGBM logloss 1.726679 trained in 7.1 seconds
39_LightGBM logloss 1.726679 trained in 7.68 seconds
40_LightGBM logloss 1.733766 trained in 5.91 seconds
41_Xgboost logloss 1.739254 trained in 10.42 seconds
42_Xgboost logloss 1.741754 trained in 8.58 seconds
* Step hill_climbing_2 will try to check up to 13 models
43_CatBoost logloss 1.56187 trained in 18.33 seconds
44_CatBoost logloss 1.579826 trained in 7.82 seconds
45_CatBoost logloss 1.544089 trained in 25.28 seconds
46_CatBoost logloss 1.573321 trained in 15.89 seconds
47_Xgboost logloss 1.636228 trained in 12.63 seconds
48_Xgboost logloss 1.739344 trained in 10.01 seconds
49_Xgboost logloss 1.658785 trained in 12.6 seconds
50_Xgboost logloss 1.750136 trained in 8.99 seconds
51_Xgboost logloss 1.691787 trained in 11.43 seconds
52_LightGBM logloss 1.726055 trained in 11.32 seconds
53_LightGBM logloss 1.726055 trained in 11.31 seconds
54_LightGBM logloss 1.732379 trained in 8.95 seconds
55_LightGBM logloss 1.742751 trained in 8.25 seconds
* Step boost_on_errors will try to check up to 1 model
31_CatBoost_BoostOnErrors logloss 1.555208 trained in 22.28 seconds
* Step ensemble will try to check up to 1 model
Ensemble logloss 1.518612 trained in 9.07 seconds
* Step stack will try to check up to 29 models
31_CatBoost_Stacked logloss 1.689158 trained in 26.0 seconds
47_Xgboost_Stacked logloss 1.712571 trained in 19.45 seconds
37_LightGBM_Stacked logloss 1.730932 trained in 9.19 seconds
45_CatBoost_Stacked logloss 1.700056 trained in 21.63 seconds
49_Xgboost_Stacked logloss 1.719332 trained in 15.79 seconds
19_LightGBM_Stacked logloss 1.730932 trained in 9.96 seconds
43_CatBoost_Stacked logloss 1.698918 trained in 20.02 seconds
11_Xgboost_Stacked logloss 1.696216 trained in 14.0 seconds
53_LightGBM_Stacked logloss 1.74761 trained in 9.76 seconds
32_CatBoost_Stacked logloss 1.700918 trained in 12.34 seconds
51_Xgboost_Stacked logloss 1.727752 trained in 15.48 seconds
52_LightGBM_Stacked logloss 1.74761 trained in 9.94 seconds
26_CatBoost_Stacked logloss 1.708754 trained in 21.03 seconds
35_Xgboost_Stacked logloss 1.694815 trained in 12.66 seconds
20_LightGBM_Stacked logloss 1.739394 trained in 9.59 seconds
33_CatBoost_Stacked logloss 1.706031 trained in 9.78 seconds
13_Xgboost_Stacked logloss 1.749491 trained in 19.44 seconds
38_LightGBM_Stacked logloss 1.739394 trained in 9.54 seconds
25_CatBoost_Stacked logloss 1.714128 trained in 17.06 seconds
36_Xgboost_Stacked logloss 1.747747 trained in 19.28 seconds
39_LightGBM_Stacked logloss 1.739394 trained in 10.66 seconds
46_CatBoost_Stacked logloss 1.70973 trained in 17.88 seconds
41_Xgboost_Stacked logloss 1.756194 trained in 20.1 seconds
54_LightGBM_Stacked logloss 1.736572 trained in 10.09 seconds
44_CatBoost_Stacked logloss 1.723194 trained in 10.77 seconds
48_Xgboost_Stacked logloss 1.675515 trained in 12.51 seconds
40_LightGBM_Stacked logloss 1.712337 trained in 9.1 seconds
42_Xgboost_Stacked logloss 1.757202 trained in 19.77 seconds
14_LightGBM_Stacked logloss 1.712337 trained in 9.29 seconds
* Step ensemble_stacked will try to check up to 1 model
Ensemble_Stacked logloss 1.518609 trained in 19.73 seconds
AutoML fit time: 1193.06 seconds
AutoML best model: Ensemble_Stacked
     prediction_0  prediction_1  prediction_2  prediction_3  prediction_4  prediction_5  prediction_6  prediction_7  label
0        0.014220      0.056234      0.165420      0.035292      0.042058      0.167955      0.435675      0.083146      6
1        0.045392      0.020603      0.017285      0.423570      0.041578      0.347478      0.070152      0.033942      3
2        0.048553      0.020185      0.058456      0.085385      0.380636      0.107843      0.046192      0.252750      4
3        0.173253      0.085424      0.023371      0.068296      0.073223      0.460410      0.041409      0.074616      5
4        0.040559      0.326146      0.039157      0.105880      0.167995      0.115706      0.110964      0.093594      1
..            ...           ...           ...           ...           ...           ...           ...           ...    ...
195      0.372883      0.319299      0.071409      0.027826      0.083929      0.043603      0.012973      0.068078      0
196      0.102378      0.020080      0.311850      0.145061      0.019909      0.125841      0.216463      0.058418      2
197      0.036133      0.270072      0.172923      0.094075      0.029849      0.026208      0.298958      0.071782      6
198      0.097706      0.014617      0.024061      0.164259      0.054482      0.485713      0.056686      0.102476      5
199      0.394162      0.046654      0.104227      0.040062      0.006349      0.361473      0.008009      0.039063      0

[200 rows x 9 columns]
     prediction_0  prediction_1  prediction_2  prediction_3  prediction_4  prediction_5  prediction_6  prediction_7  label
0        0.014220      0.056234      0.165420      0.035292      0.042058      0.167955      0.435675      0.083146      6
1        0.045392      0.020603      0.017285      0.423570      0.041578      0.347478      0.070152      0.033942      3
2        0.048553      0.020185      0.058456      0.085385      0.380636      0.107843      0.046192      0.252750      4
3        0.173253      0.085424      0.023371      0.068296      0.073223      0.460410      0.041409      0.074616      5
4        0.040559      0.326146      0.039157      0.105880      0.167995      0.115706      0.110964      0.093594      1
..            ...           ...           ...           ...           ...           ...           ...           ...    ...
195      0.372883      0.319299      0.071409      0.027826      0.083929      0.043603      0.012973      0.068078      0
196      0.102378      0.020080      0.311850      0.145061      0.019909      0.125841      0.216463      0.058418      2
197      0.036133      0.270072      0.172923      0.094075      0.029849      0.026208      0.298958      0.071782      6
198      0.097706      0.014617      0.024061      0.164259      0.054482      0.485713      0.056686      0.102476      5
199      0.394162      0.046654      0.104227      0.040062      0.006349      0.361473      0.008009      0.039063      0

[200 rows x 9 columns]
```

If you can't share your code and data publicly, maybe you can send it by email?

**Comment by alitirmizi23** [comment]
@pplonski unfortunately I can't. I'll try to replicate your dummy data code and see if that works.

**Comment by alitirmizi23** [comment]
Thanks @pplonski for the test though. I'll let you know once I figure out..
